pkg/storage/cloudimpl/aws_kms.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+package cloudimpl
+
+import (
+	"context"
+	"net/url"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/kms"
+	"github.com/cockroachdb/cockroach/pkg/storage/cloud"
+	"github.com/cockroachdb/errors"
+)
+
+const awsScheme = "aws"
+
+type awsKMS struct {
+	kms                 *kms.KMS
+	customerMasterKeyID string
+}
+
+var _ cloud.KMS = &awsKMS{}
+
+func init() {
+	cloud.RegisterKMSFromURIFactory(MakeKMSFromURI, awsScheme)
+}
+
+type kmsURIParams struct {
+	accessKey string
+	secret    string
+	tempToken string
+	endpoint  string
+	region    string
+	auth      string
+}
+
+func resolveKMSURIParams(kmsURI url.URL) kmsURIParams {
+	kmsURIParams := kmsURIParams{
+		kmsURI.Query().Get(AWSAccessKeyParam),
+		kmsURI.Query().Get(AWSSecretParam),
+		kmsURI.Query().Get(AWSTempTokenParam),
+		kmsURI.Query().Get(AWSEndpointParam),
+		kmsURI.Query().Get(KMSRegionParam),
+		kmsURI.Query().Get(AuthParam),
+	}
+
+	// AWS secrets often contain + characters, which must be escaped when
+	// included in a query string; otherwise, they represent a space character.
+	// More than a few users have been bitten by this.
+	//
+	// Luckily, AWS secrets are base64-encoded data and thus will never actually
+	// contain spaces. We can convert any space characters we see to +
+	// characters to recover the original secret.
+	kmsURIParams.secret = strings.Replace(kmsURIParams.secret, " ", "+", -1)
+	return kmsURIParams
+}
+
+// MakeKMSFromURI is the factory method which returns a configured, ready-to-use
+// AWS KMS object.
+func MakeKMSFromURI(uri string, env cloud.KMSEnv) (cloud.KMS, error) {
```

Comment:
Should this be `MakeAWSKMS` to indicate its AWS-speicific nature? Also, if all (one of) the constructors are from URI, we can probably skip "FromURI" suffix.